Refactor to fix trafficgen UIDs (back-port)

portante opened the pull request:
This is a back-port (of-sorts) of PR #2071 from `main`.  It is really a re-implementation of that PR on the `b0.69` branch using a series of commits to painstakingly fix up the `pbench-trafficgen` code to be a bit more supportable and allow it to work on this branch.

WE HAVE NO GOOD WAY TO TEST THIS FUNCTIONALITY.

Not having a sufficient platform to test this functionality has proven to make the task of maintaining this code very difficult.

The original comment from `main` is provided below.

----

The `pbench-trafficgen` script generates unique iterations based on the following command line argument (provided or defaulted):

 * `--traffic-profiles`
 * `--rates`
 * `--traffic-directions`
 * `--frame-sizes`
 * `--num-flows`
 * `--max-loss-pcts`

However, the UIDs for each generated sample was using the following set of data:

 * `-traffic_generator:%traffic_generator%-trial:%trial%`

This can lead to situations where one asks for `pbench-trafficgen` to run multiple iterations, where if the trial numbers don't match the iteration data won't be merged.

This set of modifications ensures that the UIDs used in the generated data match the values used to create each iteration.

One of the behaviors observed previously was that one could see an error like the following (long lines artificially wrapped):

```
Error: values do not match
key: claimed_dev_pairs values: ARRAY(0x559b88b50c20) \
  ARRAY(0x559b88b5f750)
Error: values do not match
key: null_stats values: HASH(0x559b88b5d230) HASH(0x559b88b5df68)
Error: values do not match
key: output_dir values: \
  /var/lib/pbench-agent/\
    trafficgen_RHOSP16.1-RHEL8.2-OVS-OFFLOAD\
      _tg:trex_r:none_fs:1500_nf:128_fm:si_td:bi_ml:0.002_tt:bs_\
      _2021-01-13T10:43:04/\
    1-bidirectional-1500B-128flows-0.002pct_drop/sample2 \
  /var/lib/pbench-agent/\
    trafficgen_RHOSP16.1-RHEL8.2-OVS-OFFLOAD\
      _tg:trex_r:none_fs:1500_nf:128_fm:si_td:bi_ml:0.002_tt:bs_\
      _2021-01-13T10:43:04/\
    1-bidirectional-1500B-128flows-0.002pct_drop/sample3
Error: values do not match
key: test_dev_pairs values: ARRAY(0x559b88b5d500) ARRAY(0x559b88b5f498)
```

This output was due in part to the UIDs not matching because the `trial` value was different.  But even if we resolve that difference, the `output_dir` key will always be different between samples, causing this error to be emitted.  Further, differences between Perl arrays and hashes were not being handled properly, causing the other errors emitted above.

This change addresses those problems as well.

ndokos reviewed (changes requested):
Not sure whether any of the commit messages will survive, but there are two typos that I saw:
- in the commit  entitled "Only use inner most loop arguments":  were --> where.
- in the commit entitled "Refactor to fix trafficgen UIDs (back-port)": pain-stakenly --> painstakingly.
The latter was also on the PR description, but I fixed it there.

I also noticed that the only test changes in the commit entitled "Only emit the benchmark exit code on failure" were deletions of the 0 exit code: there were no additions of non-zero exit codes, so there are no unit tests testing that.

Otherwise, LGTM.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -178,13 +180,43 @@ my %last_trial = %{ $$json_result{'trials'}[-1] };
+	$config{get_label('uid_label')} .= "-num_flows:MISSING";
+}
+if (exists($config{'max_loss_pct'}) && defined($config{'max_loss_pct'})) {
+	$config{get_label('uid_label')} .= "-pct_drop:%max_loss_pct%";
```
Should add missing here just as we do for the other parameters on lines 213-216.

portante commented on agent/bench-scripts/tests/pbench-user-benchmark/test-09.txt:
```diff
@@ -52,3 +52,6 @@ user_script = user-benchmark-script, no-file
 +++ pbench-user-benchmark_test-09_1900.01.01T00.00.00/./1-default/sample1/user-benchmark.cmd file contents
 user-benchmark-script no-file 2>&1 | tee /var/tmp/pbench-test-bench/pbench-agent/pbench-user-benchmark_test-09_1900.01.01T00.00.00/1-default/sample1/result.txt
 --- pbench-user-benchmark_test-09_1900.01.01T00.00.00/./1-default/sample1/user-benchmark.cmd file contents
++++ pbench-user-benchmark_test-09_1900.01.01T00.00.00/./1-default/sample1/user-benchmark.cmd file contents
```
This should not be repeated.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1151,191 @@ function inner_most_loop() {
-			if [ $traffic_generator == "trex-txrx-profile" ]; then
-				if [ "${traffic_profiles}" == "" ]; then
+				if [[ "${traffic_profile_arg}" == "" ]]; then
 					profile_builder_cmd_file="${iteration_dir}/profile-builder-cmd"
```
This should really be named `profile-builder.cmd` so that the unit tests will pick it up in the gold output.

portante commented on agent/bench-scripts/unittests:
```diff
@@ -156,6 +156,9 @@ function _dump_logs {
+        for f in $(find . -name '*.cmd' 2> /dev/null | sort); do
+            _dump_file ${1} ${f}
+        done
         for f in $(find . -name user-benchmark.cmd 2> /dev/null | sort); do
```
This is responsible for the duplicate output.

portante commented:
Thanks for the review, @ndokos.  I have updated a few of the commits based on a few of the resolved comments, and that caused a few more bugs to be exposed.  In particular, the `trafficgen-postprocess` script was incorrectly checking for a parameter value in the `binary-search.json` document.

I have also addressed all of your previous comments from your review.

portante commented:
I believe this code is sufficiently complete for what was changed.  Unfortunately, we have an example of another test run where `process-iteration-samples.out` reports errors for conflicting values:

```
Error: values do not match
key: rate
	iteration value: 50.0963504272461
	sample    value: 49.9975235470459
Error: values do not match
key: rate
	iteration value: 50.0963504272461
	sample    value: 50.0225348144531
Error: values do not match
key: trial
	iteration value: 23
	sample    value: 21
Error: values do not match
key: trial_primary_output_file
	iteration value: binary-search.trial-023.txt
	sample    value: binary-search.trial-021.txt
Error: values do not match
key: trial_secondary_output_file
	iteration value: binary-search.trial-023.extra.txt
	sample    value: binary-search.trial-021.extra.txt
```

This is really exposing an assumption the post-processing code is making that is not valid: their are fields in the `trial_params` which are outputs of the trial, not parameters of the trial: `rate`, `trial`, `trial_primary_output_file`, `trial_secondary_output_file`.

The output file names should be dropped, like we already do for the `output_dir` field, since they are in the `binary-search.json` file, and are not material to the post-processed data.  The `trial` ID and the `rate` at which the trial achieved the desired result should be store with each sample value in each of the arrays.  The original `rate_arg` should be stored as `rate` in the benchmark parameters instead.

portante reviewed (commented):
Okay, ready for review again.  Fix the problems with unwanted `flake8` being applied to `bash` scripts (they are now Python3).

portante commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -226,18 +233,27 @@ foreach $sample (@json_samples) { # samples 0..N
+							# For the rest of the keys, either check to see if it matches existing value in the %iteration,
+							# or if it does not exist, create it.
 							if ( $key ne get_label('timeseries_label') and $key ne get_label('uid_label') ) { # timeseries already handled above
 								if ($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key}) {
```
Should be using `defined` here.

portante replied:
After reading a bit, it seems like this should really be `exists()` instead of `defined()`, both here and in other places.

ndokos reviewed (commented):
Reviewed all commits except the big one: "Refactor to fix trafficgen UIDs (back-port)" (which I reviewed before).

LGTM - I'll do the big one next.

webbnh reviewed (approved):
My comments are mostly nits, basic code quality suggestions, and stylistic recommendations.  (I realize that the ones on the Perl code won't be particularly welcome.... ;-) )  I didn't find anything glaringly wrong, but I didn't dive into the code that deeply, prior to our screenshare.

I shared my thoughts on restructuring the Trafficgen data with you separately.  I suspect that pursuing them under a separate PR would probably be appropriate.

ndokos commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1151,191 @@ function inner_most_loop() {
 		fi
 
-		inner_most_loop "profile"
+		inner_most_loop "profile" "${traffic_profile}" "no-rate" "no-traffic-direction" "no-frame-size" "no-num-flow"
```
I have the same feeling.

portante replied:
I do too.

This would be good for a separate PR on the `main` branch instead.

ndokos commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -995,31 +1031,83 @@ else
 
 	for max_loss_pct in `echo $max_loss_pcts | sed -e s/,/" "/g`; do
-		if [ $rc -ne 0 ]; then
+		if [[ ${rc} -ne 0 ]]; then
```
I believe that is the global `rc` that is initialized in line 1311 in the "main" program  and is only modified by line 1235 in the `inner_loop` function.

webbnh replied:
If you change "initialized" to "re-initialized", then I agree with your statement, Nick.  :-)

portante replied:
How is this "re-initialized" at line 1311?  In the flow of `bash`, the function `inner_most_loop` is only defined at line 1062, it invoked until 1318 or 1338.  So isn't line 1311 the place where `rc` is initialized?  Or is there something we're missing here with bash behaviors?

webbnh replied:
Nope, you (both) are right -- this code is inside a function which is accessing `rc` from its caller's scope...ick.  (There is so much wrong with this that I think I'll just shut up, now.)

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -40,58 +38,73 @@ fi
-trex_use_ht="n"
-trex_use_l2="n"
+benchmark_name="trafficgen"
+export benchmark=${benchmark_name}
```
Agreed.

Unfortunately, this is a defined interface behavior of the pbench agent, which should be changed, but I can't change that now in this commit.

portante replied:
This would be a great task for you to take up as part of your ramp up on the code on the `main` branch.  What do you think?

webbnh replied:
Open an issue?

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -442,179 +456,179 @@ if [ $? -ne 0 ]; then
 	printf -- "${script_name} $*\n"
 	printf -- "\n"
 	printf -- "\tunrecognized option specified\n\n"
-	usage
+	usage >&2
```
Almost, but not quite.  The about print statements should also be changed.

portante replied:
I had apparently fixed this in the `main` branch but failed to backport it.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -915,6 +915,38 @@ fi
 ###
 
+# Build up a result directory identifier from parameters
+tg_config="tg:$(echo ${traffic_generator} | sed -e 's/-txrx//')"
+if [ $traffic_generator == "trex-txrx-profile" -a "${traffic_profiles}" != "" ]; then
+	traffic_profiles_list=""
+	for traffic_profile in $(echo ${traffic_profiles} | sed -e s/,/" "/g); do
```
Agreed.

I was trying not to make those kinds of changes because they were not directly related to fixing this problem, and those loop control constructs, if borked by an attempt to replace the external use of `sed` with `bash` equivalent, could cause more problems.

I'd rather do that kind of change outside this PR.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -995,31 +1031,83 @@ else
+		pbench-register-tool-trigger --group=$tool_group --start-trigger="Starting binary-search" --stop-trigger="Finished binary-search"
+	elif [ "$tool_period" == "repeat-final-validation" ]; then
+		pbench-register-tool-trigger --group=$tool_group --start-trigger="Starting repeat final validation for debug collection" --stop-trigger="Finished repeat final validation for debug collection"
+	fi
```
Agreed, on both points.

I just tried to lift-n-shift this code to a new location without changing the original.

But this seems worth-while to make another commit to fix that for clarity, and to add a defense against a bad `--tool-period` argument.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,54 @@ close TXT;
 $config{get_label('uid_label')} = sprintf("%s-%s", $benchmark_name, $benchmark_version);
-if (exists($config{'trial'}) && defined($config{'trial'}) && exists($config{'traffic_generator'}) && defined($config{'traffic_generator'})) {
-	$config{get_label('uid_label')} .= "-traffic_generator:%traffic_generator%-trial:%trial%";
+if (exists($config{'traffic_generator'}) && defined($config{'traffic_generator'})) {
```
I don't know why both.  I made the change to follow the previous method of handling things.  I'd rather not make deeper changes to Perl code if I can avoid it.

portante replied:
I wonder if this is a defense against `null` values in the original JSON for some reason?  I do see `null` in some of the `binary-search.json` files we have collected over time.

portante replied:
This is something we can consider addressing as a separate commit in a PR where we forward port this work to the `main` branch.

webbnh replied:
I think you're right that it's a defense against the presence of null values.  Nevertheless, having it abstracted would be nice but, as you say, is beyond the scope of this PR.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,54 @@ close TXT;
+		$config{'rate_arg'} = $rate_arg;
+		$config{get_label('uid_label')} .= "-rate:%rate_arg%";
+		if (exists($config{'traffic_direction'}) && defined($config{'traffic_direction'})) {
+			$config{get_label('uid_label')} .= "-traffic_direction:%traffic_direction%";
```
This is just raw string concatenation as far as I can tell.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,54 @@ close TXT;
+		if (exists($config{'traffic_direction'}) && defined($config{'traffic_direction'})) {
+			$config{get_label('uid_label')} .= "-traffic_direction:%traffic_direction%";
+		} else {
+			$config{get_label('uid_label')} .= "-traffic_direction:MISSING";
```
Yes, this whole thing should be refactored better.  I'll take a look again at this.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,54 @@ close TXT;
+		$config{get_label('uid_label')} .= "-num_flows:%num_flows%";
+	}
+} else {
+	$config{get_label('uid_label')} .= "-rate:MISSING";
+	$config{get_label('uid_label')} .= "-traffic_direction:MISSING";
+	$config{get_label('uid_label')} .= "-frame_size:MISSING";
+	$config{get_label('uid_label')} .= "-num_flows:MISSING";
```
Yes, this is icky, and I should not have held my nose and perpetuated it.

portante commented on agent/bench-scripts/tests/pbench-trafficgen/mock.trafficgen/binary-search.py:
```diff
@@ -0,0 +1,20 @@
+    args = " ".join(sys.argv)
+    ofp.write(f"{args}\n")
+
+val = os.environ.get("_PBENCH_UNIT_TESTS_NO_MATCH_BINARY_SEARCH", "")
+if os.environ.get("_PBENCH_UNIT_TESTS_NO_MATCH_BINARY_SEARCH", ""):
+    bad_msg = "something other than a "
+else:
+    bad_msg = val
+
+print(f"Starting {bad_msg}binary-search")
```
There is a bug here, I think those are supposed to be two different environment variables.

portante replied:
After reviewing, no bug, other than what you pointed out.  Fixed.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1151,191 @@ function inner_most_loop() {
+				if [[ "${traffic_profile_arg}" == "" ]]; then
+					profile_builder_cmd_file="${iteration_dir}/profile-builder.cmd"
+
+					echo "pushd ${trafficgen_dir} > /dev/null" > ${profile_builder_cmd_file}
+					echo "./profile-builder.py --frame-size=${frame_size_arg} --num-flows=${num_flow_arg} --rate=${rate_arg} ${profile_flow_mods} --traffic-direction=${traffic_direction_arg} --measure-latency" >> ${profile_builder_cmd_file}
+					echo "cmd_rc=\${?}" >> ${profile_builder_cmd_file}
+					echo "popd > /dev/null" >> ${profile_builder_cmd_file}
+					echo "exit \${cmd_rc}" >> ${profile_builder_cmd_file}
```
The external file is a mechanism is typically used in bench-scripts to make sure the variables are passed correctly to the external command.  However, this use of that mechanism is not taking advantage of quoting the parameters to ensure that.

So we should either:  use the mechanism correctly, or drop it in favor of a direct invocation.

Either way, I'd rather not make that change in this PR on this branch (something we should do on the `main` branch in a separate PR).

portante replied:
The other reason for these external commands is to record how various commands were invoked, as these files are stored in the result directories for later inspection.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -962,14 +996,16 @@ function record_profile_iteration {
 }
 
-mkdir -p $benchmark_run_dir/.running
+if [ "${postprocess_only}" == "n" ]; then
```
I agree with this sentiment.

I'd rather do this in a PR on the `main` branch.

We have to forward port all these changes to the `main` branch anyways, so I'll take care to make that happen.

portante commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -254,24 +270,24 @@ foreach $sample (@json_samples) { # samples 0..N
-						$metric_name eq "throughput" and
-						$metric_type eq $primary_metric and
-						$iteration{$metric_name}{$metric_type}[$i]{get_label('stddevpct_label')} > $max_stddevpct) {
+					($iteration{$metric_type}{$metric_name}[$i]{get_label('mean_label')},
```
This is something we should tackle in a PR on the `main` branch when we move this forward there.

dbutenhof replied:
Yeah, seems like `$ir = \$iteration{$metric_type}{$metric_name}[$]` would help a lot with readability. Or maybe it has to be `$ir = \{...}`?

I'm not sure I can ever forgive you for making my Perl start to come back, Pete ... can we just get this done so I can sink back into blissful forgetfulness?

portante replied:
I'd rather not touch this code any further in this PR.  We can certainly fix this up on the `main` branch, if we even decide to forward port this PR to `main`.

Here, all I wanted to do is correct the inconsistency: lines 188 and 196 use `$metric_type` and then `$metric_name`, but this nested loop had them reversed.

portante commented:
I have addressed all the code review comments that I felt were safe to do on this branch.  I think the other comments are worth doing when we forward port this to the the `main` branch.

All code review comments have been addressed in new commits, so the old comments should still be visible at the point where they were made.  All comments that have been addressed in the commits have been marked as resolved.

riya-17 reviewed (dismissed):
Looks Good

dbutenhof commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -220,24 +227,33 @@ foreach $sample (@json_samples) { # samples 0..N
+								if (exists($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key})) {
 									my $iteration_value = $iteration{$metric_type}{$metric_name}[$iteration_matching_index]{$key};
 									if ("$sample_value" ne "$iteration_value") {
-										print "Error: values do not match\nkey: $key values: $iteration_value $sample_value\n";
+										my $sv = ref($sample_value);
+										my $iv = ref($iteration_value);
+										if ($sv and ($sv eq $iv)) {
+											my $sv_s = Dumper($sample_value);
+											my $iv_s = Dumper($iteration_value);
+											if ($sv_s ne $iv_s) {
+												print "Error: values do not match\n\tkey: $key\n\titeration value: $iv_s\n\tsample    value: $sv_s\n";
+											}
+										} else {
+											print "Error: values do not match\n\tkey: $key\n\titeration value: $iteration_value\n\tsample    value: $sample_value\n";
+										}
```
I am __so__ confused. Trying to read Perl gives me such a headache.

So, if a Perl comparison of `$sample_value` and `$iteration_value` shows they're _not_ equal, we get `ref()` of both to see whether they are refs to the same type; if so, then we `Dumper` both to get a deep representation of whatever they reference and compare _those_; we fail if the `Dumper`s don't match, and we fail if they're not identical types in the first place.

Why do they have to be identical objects just because they're the same type? And a better error message for the `else` would be that they're not the same type rather than repeating that the values don't match.

dbutenhof replied:
> > I am **so** confused.
> 
> Based on what you wrote below, you do not appear to be confused. =)

I was quite sure I was confused; so if I'm not, that's confusing...

> > Why do they have to be identical objects just because they're the same type?
> 
> In the first place, the original code was emitting error messages when the iteration key and the sample key have different values.

Right.

> In the case where the values are not scalars, they would always emit the error message, but an array and a hash _always_ have different scalar values.

I don't think you explained that clearly, but I get it: two identical hashes or two identical arrays (which don't happen to be refs to the same Perl object) would have appeared different before but now will be diagnosed as equal.

> This change is trying to fix that. If each key is referencing the same type, and the `Dumper` of each are identical, then we consider them to have the same data, and we don't emit an error message.
> 
> If they are not the same type, they certainly can't be the same values, so we just emit the same message as before.
> 
> > And a better error message for the `else` would be that they're not the same type rather than repeating that the values don't match.
> 
> Well, that doesn't help the user. They don't know anything about Perl. That just want to know what keys are being compared and why the don't match. So this code only emits information about the keys and values.

I always find explanations like this a bit concerning. Basically, you _are_ telling them that the Perl types don't match, because the implementation is in Perl and the semantics aren't separable. Then again, I can't immediately construct a case where the two strings you dump would "appear equivalent" but be distinct to Perl so I'm not going to pursue that debate; I don't think it really matters, although I thought it might be useful to _us_ when someone shows us the error message.

portante replied:
It is pretty easy to see the difference: when you have a hash or an array it is displayed as such, vs a string, float, or integer which don't look like those objects as rendered by `Dumper`.

webbnh commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1151,191 @@ function inner_most_loop() {
-				done
+					echo_log "Not going to run $benchmark_name.  Only postprocesing existing data"
+				fi
 				if [ $rc -eq 0 ]; then
```
Nope...as pointed out above, `rc` is initialized in the caller's scope before this code (which is a function) is invoked.

ndokos commented:
The postrprocess-iteration-samples-0 unit test fails.

ndokos commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -220,24 +227,33 @@ foreach $sample (@json_samples) { # samples 0..N
 								$samples_hash{get_label('timeseries_label')} = $sample_hash{$metric_type}{$metric_name}[$i]{get_label('timeseries_label')};
 							}
-							if ( not defined $iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')} ) {
+							if (exists($iteration{$metric_type}{$metric_name}[$iteration_matching_index]{get_label('samples_label')})) {
```
I don't know if `exists(foo)` and `foo` are equivalent in an `if`, but I was assuming that they are and that you just added `exists` everywhere where that pattern held. But this change does not follow the pattern and seems broken.

portante replied:
Good catch.  This is incorrect and simply adding a `not` to restore the proper test works to restore the unit tests to a passing state.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1151,191 @@ function inner_most_loop() {
-				else
-					error_log "Aborting benchmark"
-					break
+					echo "${script_path}/postprocess/${benchmark_name}-postprocess \"${benchmark_results_dir}\" \"${full_hostname}\" \"${rate_arg}\" \"${frame_size_arg}\" \"${num_flow_arg}\" \"${traffic_profile_name}\" \"${trafficgen_version}\"" > "${benchmark_results_dir}/${benchmark_name}-postprocess.cmd"
+					chmod +x "${benchmark_results_dir}/${benchmark_name}-postprocess.cmd"
+					${benchmark_results_dir}/${benchmark_name}-postprocess.cmd 2>&1 | tee ${benchmark_results_dir}/${benchmark_name}-postprocess.out || rc=1
```
The external file is a mechanism is typically used in bench-scripts to make sure the variables are passed correctly to the external command. However, this use of that mechanism is not taking advantage of quoting the parameters to ensure that.

The other reason for these external commands is to record how various commands were invoked, as these files are stored in the result directories for later inspection.

portante commented:
The unit tests now run locally correctly for me, so they _should_ pass in the TravisCI environment.

I've updated commit #1266a24 with the simple fix of adding a `not` to the errant expression.

However, we still need to refactor and address the errors that will be encountered when multiple samples are used and the `rate` and `trial` output values in `binary-search.json` don't match between samples.

We could consider merging this as is, because it does fix the UIDs properly, but we'd need another PR to fix this `binary-search.json` handling.

webbnh commented:
As long as this PR doesn't make things worse in any apparent way, I'm good with merging it in.

dbutenhof commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -254,24 +270,24 @@ foreach $sample (@json_samples) { # samples 0..N
-foreach my $metric_name (sort keys %iteration) {
-	my $metric_type;
-	foreach $metric_type (sort keys %{ $iteration{$metric_name} }) {
+foreach my $metric_type (sort keys %iteration) {
+	my $metric_name;
+	foreach $metric_name (sort keys %{ $iteration{$metric_type} }) {
```
Since neither `$metric_type` nor `$metric_name` are used outside the respective loops (I think?), why are you defining the scope of `$metric_type` to the loop but letting the inner loop's `$metric_name` scope extend to the outer loop? Is there a reason?

And (I hate to ask) was there a reason for swapping the loops?

portante replied:
See line 188 and 196.

dbutenhof replied:
OK, so same iteration order; that's fine. The 188/196 loops scope both iteration variables outside the loop, completely unnecessarily; you're doing it both ways. It doesn't really impact the code, but reading the two styles in close succession makes a reader wonder why they're different; making them the same would remove that concern.

portante replied:
> OK, so same iteration order; that's fine. The 188/196 loops scope both iteration variables outside the loop, completely unnecessarily; you're doing it both ways.

What do you mean by "you're doing it both ways"?  Is the change being made here something that I am introducing that is a new behavior where I am having this loop behave differently?

At least my intention was to only change the order of the use of names.  Perhaps another commit could change lines 188/196 when we migrate this work to the `main` branch?

dbutenhof replied:
> What do you mean by "you're doing it both ways"? Is the change being made here something that I am introducing that is a new behavior where I am having this loop behave differently?

Never mind. One iteration variable is scoped inside the loop (`foreach my $metric_type`) and the other is scoped outside the loop (`my $metric_name; foreach $metric_name`) ... it makes absolutely no difference in this code pattern, it just looks odd.

webbnh replied:
I believe that Dave means that in some cases the code is declaring the loop index before/outside the loop construct while in other cases it's declaring the loop index in the loop construct.  Since the latter provides encapsulation of the variable, it is generally preferable, as long as the value isn't needed after the loop completes (it's preferable because, as readers, we know we can ignore the variable once we're past the loop).

I don't think you actually made any pertinent changes to this structure...it's just that the inconsistency is highlighted in green because you swapped the variable names.  It sounds like we have a follow-on PR coming which is going to address a number of these code quality issues....

webbnh reviewed (changes requested):
I've looked a little more into the structure of the data and the postprocessing scripts, and I've made some suggestions.  And, I've got the usual handful of nits.

portante commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -188,13 +195,13 @@ foreach $sample (@json_samples) { # samples 0..N
-							if ( $iteration{$metric_type}{$metric_name}[$j]{get_label('uid_label')} )  {
+					if (exists($iteration{$metric_type}{$metric_name})) {
+						for (my $j = 0; $j < scalar @{ $iteration{$metric_type}{$metric_name} }; $j++) {
+							if (exists($iteration{$metric_type}{$metric_name}[$j]{get_label('uid_label')}))  {
```
I don't know, I believe that historically we wanted to describe how the UID is generated, so that it gets generated the same way each time it is needed, and I think the intention was to avoid having a UID that is generated incorrectly where the description for how it is generated doesn't match.

I have never been in favor of this method as it is compute intensive (on the pbench-server side, we generate the UID once, letting the code hold the description of how it is generated).

This is not something that I would want to change as part of this PR.  If we really want to change this, we should file an issue.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1152,191 @@ function inner_most_loop() {
-							msg="tool triggers did not fire for iteration/sample, '${iteration}/sample${sample}'"
-							echo "${msg}"; log "${msg}"
+					if [[ -e "${tt_benchmark_results_dir}/tools-${tool_group}" ]]; then
+						# Post-process the collected tool data now
+						pbench-postprocess-tools --group=${tool_group} --dir=${tt_benchmark_results_dir}
```
Yes, so that any collected tool data is properly post-processed to help debug what might have failed.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1152,191 @@ function inner_most_loop() {
-	for rate in `echo $rates | sed -e s/,/" "/g`; do
-		if [ $rc -ne 0 ]; then
+	for rate in `echo ${rates} | sed -e s/,/" "/g`; do
+		if [[ ${rc} -ne 0 ]]; then
 			break
 		fi
```
I agree with moving it, perhaps when we migrate this code to the `main` branch?  I was trying not to make structural changes to the code where possible.  Most of the restructuring I took on was so that I could follow the logic of the code better, since it was pretty convoluted.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1152,191 @@ function inner_most_loop() {
-		if [ $rc -ne 0 ]; then
+if [[ ${traffic_generator} == "trex-txrx-profile" && "${traffic_profiles}" != "" ]]; then
+	for traffic_profile in `echo ${traffic_profiles} | sed -e s/,/" "/g`; do
+		if [[ ${rc} -ne 0 ]]; then
 			break
 		fi
```
I agree with moving it, perhaps when we migrate this code to the main branch?

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -29,81 +28,78 @@ my $script_name = basename($0);
+my $frame_size_arg = $ARGV[3];
+my $num_flows_arg = $ARGV[4];
+my $traffic_profile_arg = $ARGV[5];
+my $benchmark_version = $ARGV[6];
```
I don't know enough about trafficgen to answer what it will generate.  By "pbench" I think you mean the `pbench-trafficgen` script, which offers a `--one-shot` parameter (see [line 237](https://github.com/portante/pbench/blob/go-back-fix-trafficgen/agent/bench-scripts/pbench-trafficgen#L237)).  So then the question becomes are there cases where people use `pbench-trafficgen` with `--one-shot`, and I don't know the answer to that question immediately.

webbnh replied:
Apparently, [there _are_](https://docs.google.com/document/d/1FokNXV8ZBAzac78fUoqdUBWXTxDWIEBP2YjWNpRioT8/edit#heading=h.6c70xeopatiu) such cases.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,59 @@ close TXT;
 }
+$config{get_label('uid_label')} = $uid_label;
+
 push(@benchmark, \%config);
```
Agreed.

portante commented on agent/bench-scripts/postprocess/trafficgen-postprocess:
```diff
@@ -174,25 +171,59 @@ close TXT;
-#exit;
-
 foreach my $trial_param ( keys %{ $last_trial{'trial_params'} } ) {
-	if ( $last_trial{'trial_params'}{$trial_param} ) {
+	if ( $trial_param eq 'output_dir' ) {
+		# The output_dir parameter is different per sample; we explicitly omit it
+		# so that when process-iteration-samples runs it will properly merge the
+		# same sample data sets together.
+		next;
+	}
+	if (defined($last_trial{'trial_params'}{$trial_param})) {
 		$config{$trial_param} = $last_trial{'trial_params'}{$trial_param}
 	}
 }
```
I'll noodle about this in a separate PR or as an additional commit here; not sure which one yet.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1063,214 +1152,191 @@ function inner_most_loop() {
-						chmod +x "$benchmark_results_dir/$benchmark_name-postprocess.cmd"
-						$benchmark_results_dir/$benchmark_name-postprocess.cmd 2>&1 | tee $benchmark_results_dir/$benchmark_name-postprocess.out || rc=1
+				else
+					if [[ ! -d $benchmark_results_dir ]]; then
+						error_log "Results directory $benchmark_results_dir does not exist, skipping post-processing"
+						continue
 					fi
-				done
+					echo_log "Not going to run $benchmark_name.  Only postprocesing existing data"
+				fi
```
Lines 1227 - 1267 are the `if / else / fi` for post-processing-only or not.  At line 1228 we create the `${benchmark_results_dir}` (which arguably should not be `mkdir -p`, but `mkdir` with a check for failure).  At line 1262 we just check for a missing `${benchmark_results_dir}` so that we can skip the benchmark post-processing step.  At line 1266 we only want to indicate that we didn't run the benchmark, but will post-process the benchmark data below at line 1271.

webbnh replied:
I must have missed the `continue` statement when I made this comment.  (This code is, sadly, not easy to deal with.)

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -995,31 +1037,78 @@ else
+	pbench-collect-sysinfo --group=$tool_group --dir=$benchmark_run_dir --sysinfo=$sysinfo beg
+	pbench-metadata-log --group=$tool_group --dir=$benchmark_run_dir beg
+	# on abnormal exit, make sure that the metadata log exists and is complete.
+	trap "pbench-metadata-log --group=$tool_group --dir=$benchmark_run_dir int" INT QUIT
```
Fixed.

portante commented on agent/bench-scripts/postprocess/process-iteration-samples:
```diff
@@ -220,24 +227,33 @@ foreach $sample (@json_samples) { # samples 0..N
-										print "Error: values do not match\nkey: $key values: $iteration_value $sample_value\n";
+										my $sv = ref($sample_value);
+										my $iv = ref($iteration_value);
+										if ($sv and ($sv eq $iv)) {
```
Fixed.
